kn source apiserver create/delete  (as typed-builtin source)

daisy-ycguo opened the pull request:
This is a second builtin-typed source as described in [specification](https://docs.google.com/document/d/1B8VK-5nuxt3D-i1Up3onmUbzcztQXE0GRgf1lfx49P8/edit#bookmark=id.d382ob8kwx0f), in addition to the cron-job source ib #357  

Partially address #499

## Proposed Changes

* Add command to create ApiServerSource
* Add command delete ApiServerSource

navidshaikh commented on pkg/kn/commands/source/apiserver/apiserver_flags.go:
```diff
@@ -0,0 +1,67 @@
+// Copyright © 2019 The Knative Authors
```
can we move the flags to `pkg/kn/commands/flags/` ?

daisy-ycguo replied:
These are flags for `kn source create apiserver` only.
I think it's better to keep it in `commands/source/apiserver` folder.

navidshaikh replied:
It's more about having the kn flags together in flags package, right now not all commands flags are following it (only traffic flags). We've an open issue for this https://github.com/knative/client/issues/337.

daisy-ycguo replied:
I have moved flags to that folder.

navidshaikh commented on pkg/kn/commands/source/sink/sink_flags.go:
```diff
@@ -0,0 +1,53 @@
+// Copyright © 2019 The Knative Authors
```
can we move the flags to `pkg/kn/commands/flags/` ?

daisy-ycguo replied:
Maybe sink could be moved to `pkg/kn/commands/flags/` because it will be used in several Eventing command. I will leave it to @rhuss to decide.

daisy-ycguo replied:
I have moved flags to that folder.

navidshaikh commented on pkg/kn/commands/source/source_flags.go:
```diff
@@ -0,0 +1,15 @@
+// Copyright © 2019 The Knative Authors
```
can we move the flags to `pkg/kn/commands/flags/` ?

daisy-ycguo replied:
I will suggest we keep source flags only in `kn/commands/source` folder, because it is `source` specific.

daisy-ycguo replied:
I moved all sources flags to pkg/kn/commands/flags/

toVersus commented:
@rhuss As described below, you (co-author of this PR) need to comment `@googlebot I consent`.

> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent`. in this pull request.

maximilien commented on docs/cmd/kn_source_api-server_create.md:
````diff
@@ -0,0 +1,44 @@
+### Options
+
+```
+      --accountName string   Name of the ServiceAccount to use to run this source
````
Inconsistent use of camel case between `accountName` and `apiversion`.

Unless there is a strong reason, I'd suggest to be consistent and use camel case in all cases.

daisy-ycguo replied:
done

maximilien commented on pkg/kn/commands/source/apiserver/apiserver_create.go:
```diff
@@ -0,0 +1,72 @@
+	var sinkFlags sink.SinkFlags
+
+	cmd := &cobra.Command{
+		Use:   "create NAME --resources RESOURCE --accountname ACCOUNTNAME --sink SINK --mode MODE",
```
`--accountName` ?

rhuss replied:
whyt not just `--account` ?

rhuss replied:
`--resource` instead of `--resources` ?

daisy-ycguo replied:
I plan to use `--resources` to describe a list of resources because ApiServerSource supports to define the monitoring of multiple resources.

maximilien commented on pkg/kn/commands/source/apiserver/apiserver_create.go:
```diff
@@ -0,0 +1,72 @@
+			if err != nil {
+				return err
+			}
+			aipServerSourceClient, err := NewEventingClient(config, namespace)
```
Where is the check for `err` after line 57? What if the `NewEventingClient` fails? If it can never fail then don't have it return an error.

maximilien replied:
Please address ^^^

daisy-ycguo replied:
done

maximilien commented on pkg/kn/commands/source/apiserver/apiserver_create.go:
```diff
@@ -0,0 +1,72 @@
+			if err != nil {
+				return err
+			}
+			return aipServerSourceClient.CreateApiServerSource(name, apiServerUpdateFlags.getApiServerResourceArray(), apiServerUpdateFlags.accountName, apiServerUpdateFlags.mode, objectRef)
```
This reads odd and could be improve a bit by unpacking the return. For instance:

```go
resourceArray := apiServerUpdateFlags.getApiServerResourceArray()
serverSource := aipServerSourceClient.CreateApiServerSource(name, resourceArray, apiServerUpdateFlags.accountName, apiServerUpdateFlags.mode, objectRef)
return serverSource
```

Also can any of these `CreateApiXxx(...)` call produce errors?

daisy-ycguo replied:
CreateApiServerSource could produce error. In the current code base, it has been changed to `return aipServerSourceClient.CreateApiServerSource(...)`. So the error will be returned if it is not nil.

maximilien commented on pkg/kn/commands/source/apiserver/apiserver_flags.go:
```diff
@@ -0,0 +1,67 @@
+	if len(parts) == 3 {
+		version = parts[0]
+		kind = parts[1]
+		isController, _ = strconv.ParseBool(parts[2])
```
Why are we ignoring errors? [`strconv.ParseBool`](https://golang.org/pkg/strconv/#ParseBool) returns an error if the string passed is not a bool. Should deal with error.

maximilien replied:
Please address ^^^

daisy-ycguo replied:
done

maximilien commented on pkg/kn/commands/source/source.go:
```diff
@@ -0,0 +1,32 @@
+		Use:   "source",
+		Short: "Event Source command group",
+	}
+	//sourceCmd.AddCommand(cron.NewCronCommand(p))
```
Don't leave out commented code. Just remove it if not needed.

daisy-ycguo replied:
done

daisy-ycguo commented:
Thanks for all the comments. Working on them now.

daisy-ycguo commented:
@googlebot I consent

rhuss commented:
@googlebot I consent

daisy-ycguo commented:
Thanks for comments. I will adopt them. Just add a few explanations. 

- `--mode`, mode is described [here](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go#L90) in Eventing [apiserver_types.go](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go). I don't quite get the point, just copying the explanation here. Maybe you can help to make a better description of `mode`.
- `--resources`, here is the [definition](https://github.com/knative/eventing/blob/master/pkg/apis/sources/v1alpha1/apiserver_types.go#L106) of resources in Eventing apiserver_types.go. Again, `Controller` is a boolean and the description is not clear here. I changed it to `isController` just hoping to give a clear hint that the type is boolean. `--resources` is used to describe the K8s resources that you want to monitor to. Of course users can monitor multiple resources all together. I plan to use something like `Event\v1\true,Pod\v1\false` to describe resources.

ApiServerResource is complicated than cronjob and ContainerSource. Welcome to any suggestions to get a better UX. Here is a [sample](https://knative.dev/docs/eventing/samples/kubernetes-event-source/) of ApiServerSource, which can help to get a better understanding.

daisy-ycguo commented on pkg/kn/commands/source/apiserver/apiserver_client.go:
```diff
@@ -0,0 +1,64 @@
+// Copyright © 2019 The Knative Authors
```
done

daisy-ycguo commented on pkg/kn/commands/source/apiserver/apiserver.go:
```diff
@@ -0,0 +1,30 @@
+func NewApiServerCommand(p *commands.KnParams) *cobra.Command {
+	apiServerImporterCmd := &cobra.Command{
+		Use:   "api-server",
+		Short: "Kubernetes Api Server Event Source command group",
```
done

daisy-ycguo commented on hack/build.sh:
```diff
@@ -263,13 +263,9 @@ has_flag() {
-  echo "⚔️ ${S}Compile"
+build_cross() {
+  local ld_flags="$(build_flags $(dirname $0)/..)"
+  local pkg="knative.dev/client/pkg/kn/commands"
```
done

daisy-ycguo commented:
I'm working on the testings and will add more tests in next commit.

navidshaikh commented on pkg/kn/commands/types_test.go:
```diff
@@ -141,3 +141,53 @@ func TestGetClientConfig(t *testing.T) {
+			LogHTTP:      tc.logHttp,
+		}
+
+		sourcesClient, err := p.newSourcesClient(namespace)
```
follow up suggestion based on above comment

daisy-ycguo replied:
+1

daisy-ycguo commented on pkg/kn/commands/types.go:
```diff
@@ -40,10 +42,11 @@ type Config struct {
+	KubeCfgPath      string
+	ClientConfig     clientcmd.ClientConfig
+	NewClient        func(namespace string) (serving_kn_v1alpha1.KnServingClient, error)
+	NewSourcesClient func(namespace string) (sources_kn_v1alpha1.KnSourcesClient, error)
```
agree. Thanks.

daisy-ycguo commented:
I split the common codes into a separate PR #514, for easy collaboration among all sources client features.

navidshaikh commented:
e2e flake should be fixed by https://github.com/knative/client/pull/523

daisy-ycguo commented:
/assign @rhuss

rhuss commented:
This PR is on my bucket list for today. It's a big list :)  but for sure I will review the PR until tomorrow.

navidshaikh commented:
Ran the PR locally, few things to note:
```
 ./kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc

services.serving.knative.dev "mysvc" not found
```
 - (optional) We can improve this message, but it's not consistent even between list or describe

```
➜  client git:(apis) kubectl get apiserversource k8sevents 
NAME        READY   AGE
k8sevents   False   8m21s

➜  client git:(apis) ./kn source apiserver create k8sevents --resources Event:v1:true  --sink svc:hello

ApiServerSource 'k8sevents' successfully created in namespace 'default'.
apiserversources.sources.eventing.knative.dev "k8sevents" already exists
```
 - Creation error message is printed before existence check could return.

```
./kn source apiserver create --help
[..]
      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")
```
 - We should rephrase this to "Comma separated Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")"  ?

navidshaikh commented on pkg/kn/commands/source/apiserver/create.go:
```diff
@@ -0,0 +1,103 @@
+			apisrvsrc.Spec.Sink = objectRef
+
+			// create
+			_, err = sourcesClient.CreateApiServerSource(apisrvsrc)
```
error check missing

daisy-ycguo replied:
fixed

navidshaikh commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,73 @@
+	cmd.Flags().StringVar(&f.ServiceAccountName, "service-account", "", "Name of the ServiceAccount to use to run this source")
+	cmd.Flags().StringVar(&f.Mode, "mode", "Ref", "'Ref' or 'Resource', the receive adapter controller runs under.")
+	cmd.Flags().StringVar(&f.ApiVersion, "apiVersion", "1.0", "Kubernetes API version of the resource to watch")
+	cmd.Flags().StringVar(&f.Resources, "resources", "r", "comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true")
```
should add an example like documented in code `Event:v1:true,Pod:v2:false`

daisy-ycguo replied:
done

navidshaikh commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,73 @@
+	sources_v1alpha1 "knative.dev/eventing/pkg/apis/sources/v1alpha1"
+)
+
+const ApiVersionSplitStr = ":"
```
`APIVersionSplitChar`

daisy-ycguo replied:
done

navidshaikh commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,73 @@
+)
+
+const ApiVersionSplitStr = ":"
+const ApiServerSourceArraySplitStr = ","
```
*APIServerSourceArraySplitChar

daisy-ycguo replied:
done

navidshaikh commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,73 @@
+	var isController bool
+	var err error
+	parts := strings.Split(resource, ApiVersionSplitStr)
+	if len(parts) == 3 {
```
should we check for other cases of len as 2  or 1 ?

daisy-ycguo replied:
use default value to set Version and isController, in order to support other cases as 2 or 1.

daisy-ycguo commented:
Thanks for review. Here are the updates in the latest version.

> Ran the PR locally, few things to note:
> 
> ```
>  ./kn source apiserver create k8sevents --resources Event:v1:true --service-account myaccountname --sink svc:mysvc
> 
> services.serving.knative.dev "mysvc" not found
> ```
> 
> * (optional) We can improve this message (though the error message is not consistent even between list or describe)
Updated to:
```
$ ./kn source apiserver create k8sevents --resources Event:v1:true --service-account service-sa --sink svc:mysvc
cannot create ApiServerSource 'k8sevents' in namespace 'default' because services.serving.knative.dev "mysvc" not found
```
> 
> ```
> ➜  client git:(apis) kubectl get apiserversource k8sevents 
> NAME        READY   AGE
> k8sevents   False   8m21s
> 
> ➜  client git:(apis) ./kn source apiserver create k8sevents --resources Event:v1:true  --sink svc:hello
> 
> ApiServerSource 'k8sevents' successfully created in namespace 'default'.
> apiserversources.sources.eventing.knative.dev "k8sevents" already exists
> ```
> * Creation error message is printed before existence check could return.

Updated:
```
$ k get apiserversource
NAME             AGE
k8sevents        5m15s
$ ./kn source apiserver create k8sevents --resources Event:v1:true --service-account service-sa --sink svc:fib-knative
cannot create ApiServerSource 'k8sevents' in namespace 'default' because apiserversources.sources.eventing.knative.dev "k8sevents" already exists
```
> 
> ```
> ./kn source apiserver create --help
> [..]
>       --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")
> ```
> 
> * We should rephrase this to "Comma separated Kind:APIVersion:isController list, e.g. Event:v1:true (default "r")"  ?
Updated:
```
$ ./kn source apiserver create --help
Create a Kubernetes API Server event source

Usage:
  kn source apiserver create NAME --resources RESOURCE --service-account ACCOUNTNAME --sink SINK --mode MODE [flags]

......
Flags:
......
      --resources string         comma seperate
                                 Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
```

rhuss commented on pkg/eventing/sources/v1alpha1/client.go:
```diff
@@ -23,6 +27,10 @@ import (
 type KnSourcesClient interface {
 	// Namespace in which this client is operating for
 	Namespace() string
+
```
For building sources, I wonder whether we should add an intermediate layer (like having an `KnApiServerSourceClient` which can be accessed from this interface). This would easily help in separation of concerns (as it this specialized client could be stored and picked up separately by the using code). This of course depends also on the amount of sources we expect, but even for 3 sources, I think this would be helpful (this is also how the generated service client does it).

I'm totally fine to keep it like this for the current PR but would probably refactor in a subsequent PR, when we e.. introduce the CronJobSource.

daisy-ycguo replied:
Thanks.

rhuss commented on pkg/kn/commands/flags/sink.go:
```diff
@@ -0,0 +1,58 @@
+		return nil, nil
+	}
+
+	if strings.HasPrefix(i.sink, "svc:") {
```
Good start, we will need to add more sink types (as specified) later. I'm currently working on it as part of the CronJob PR, so totally fine with it here like it is.

daisy-ycguo replied:
Agree.

rhuss commented on docs/cmd/kn_source_apiserver_create.md:
```diff
@@ -0,0 +1,43 @@
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
+  -n, --namespace string         Specify the namespace to operate in.
+      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
+      --service-account string   Name of the ServiceAccount to use to run this source
```
```suggestion
      --service-account string   Name of the service account to use for running this source
```

(for aliging with `kn service` commands

daisy-ycguo replied:
fixed in create_flag.go

rhuss commented on pkg/kn/commands/source/apiserver/create.go:
```diff
@@ -0,0 +1,110 @@
+			if err != nil {
+				return fmt.Errorf(
+					"cannot create ApiServerSource '%s' in namespace '%s' "+
+						"because %s", name, namespace, err)
```
You'd better use `%v` instead of `%s` when formatting an error.

rhuss commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func getValidResource(resource string) (string, string, bool) {
+	var version = DefaultApiVersion // v1 as default
```
would call it `apiGroupVersion` as this is supposed be the api-group + version combined (like in in `apps/v1` or `serving.knative.dev/v1alpha`)

rhuss replied:
Also, IMO the version should be mandatory. Having `v1` as a default doesn't not make sense except for the most basice objects which does not carry an API group.

rhuss commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,77 @@
+	Resources          string
+}
+
+// GetApiServerResourceArray is to return an array of ApiServerResource from a string. A sample is Event:v1:true,Pod:v2:false
```
Please don't use `Pod:v2` as an example as this does not exist. Instead use also an example with an apiGroup (like in `Deployment:apps/v1`)

rhuss commented on pkg/kn/commands/source/apiserver/create_flag.go:
```diff
@@ -0,0 +1,77 @@
+// GetApiServerResourceArray is to return an array of ApiServerResource from a string. A sample is Event:v1:true,Pod:v2:false
+func (f *ApiServerSourceUpdateFlags) GetApiServerResourceArray() []sources_v1alpha1.ApiServerResource {
+	var resourceList []sources_v1alpha1.ApiServerResource
+	for _, r := range strings.Split(f.Resources, ApiServerSourceArraySplitChar) {
```
I'm not sure if already had that discussion, but I would much prefer to allow multiple options for `--resource` instead of cramming everything into a single one, comma separated. It should be modelled much like `--traffic` on `kn service update`, to be consistent and and e.g. also for clarity.

daisy-ycguo replied:
`--traffic` is `StringSliceVar`, for example `v1=50,v2=50`; I will follow `--traffic` to handle `--resource`.

rhuss commented on docs/cmd/kn_source_apiserver_create.md:
```diff
@@ -0,0 +1,43 @@
+  -h, --help                     help for create
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
+  -n, --namespace string         Specify the namespace to operate in.
+      --resources string         comma seperate Kind:APIVersion:isController list, e.g. Event:v1:true,Pod:v2:false
```
Please use a realistic example and maybe avoid the comma separated approach. See my comments below please.

rhuss replied:
What does the boolean `controller` flag mean? Can we add this here into the documentation? Also please add which parts are mandatory and which are optional.

If this is too long then add the explanation outside of the options to the top-level command doc.

daisy-ycguo replied:
done

rhuss commented on docs/cmd/kn_source_apiserver_create.md:
````diff
@@ -0,0 +1,43 @@
+
+```
+  -h, --help                     help for create
+      --mode string              'Ref' or 'Resource', the receive adapter controller runs under. (default "Ref")
````
Could you briefly explain here the difference between `Ref` and `Resource` ?

daisy-ycguo replied:
done.

rhuss commented on docs/cmd/kn_source_apiserver_create.md:
```diff
@@ -0,0 +1,43 @@
+### Synopsis
+
+Create a Kubernetes API Server event source
+
```
Here you could explain a bit the different modes and what a controller is, which kinds of events are delivered (== all), etc.

daisy-ycguo replied:
done

rhuss commented on pkg/kn/commands/source/apiserver/delete.go:
```diff
@@ -0,0 +1,59 @@
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "ApiServerSource '%s' successfully deleted in namespace '%s'.\n", args[0], namespace)
```
```suggestion
			fmt.Fprintf(cmd.OutOrStdout(), "ApiServerSource '%s' deleted in namespace '%s'.\n", args[0], namespace)
```

daisy-ycguo replied:
fixed

daisy-ycguo commented on pkg/eventing/sources/v1alpha1/client.go:
```diff
@@ -23,6 +27,10 @@ import (
 	// Namespace in which this client is operating for
 	Namespace() string
+
+	CreateApiServerSource(apisvrsrc *v1alpha1.ApiServerSource) (*v1alpha1.ApiServerSource, error)
```
addressed.

daisy-ycguo commented on pkg/eventing/sources/v1alpha1/client_test.go:
```diff
@@ -0,0 +1,108 @@
+
+	apisourceNew := newApiServerSource(srcName)
+
+	sourcesServer.AddReactor("create", "apiserversources",
```
Thanks. I will create an issue to track.

rhuss replied:
On a second thought having a dedicated mock implementation introduced only for the source looks like overkill. In fact for the CronJob source I even went so far to make it package private to reduce coupling (i.e. do not introduce the client into the KnCommand). Still fiddling for how to get the fake client injected into the test, but I think the more self-contained we get the source implementations (without having top-level hooks like a global KnSourcesClient), the better, e.g. for latter refactoring (e.g. when the ApiServer or CronJob sources will move out of eventing which is not unlikely).

daisy-ycguo commented:
@googlebot I consent.

rhuss commented:
> @googlebot I consent.

Hmm, seems not to be picked up. wonder whether this is a bot issue or more general.

daisy-ycguo commented:
@rhuss The major changes in the last version:
- follow `--traffic` to manage flag `--resource`
- add enough documentation to explain the flags
- minor changes.

Please review again. Thank you.

rhuss commented:
@daisy-ycguo thanks ! There are some minor things to fix, but let's save a round-trip an merge now. If you don't mind I would update the minor fixes (like doc wording etc) in an extra PR.

/lgtm
/approve
